RD-1607 Change missing secrets error in Deploy Blueprints

leon-good-life opened the pull request:
<!--- Provide JIRA issue ID and a general summary of your changes in the Title above -->

## Description
This pull request parses the Missing Secrets Error Message and shows the error correctly. And allows the user to add missing secrets.
<!--- Describe your changes to help reviewers understand the details. -->

## Screenshots / Videos
**Updated video:**

https://user-images.githubusercontent.com/17104212/145029262-64d8374a-3dd1-4bdf-8447-3150537953c8.mov

<!--- If applicable, add screenshot(s) or video(s) describing the changes you made. -->
<!--- Consider adding comparison screenshot(s) - before and after the change. -->
<!--- If not applicable, just write “N/A” in this section. -->

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] My code follows the UI Code Style.
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires a documentation update.
- [x] I added proper labels to this PR.

## Tests

### System tests
* I added a test that uploads a blueprint with required secrets, and attempting to deploy and receiving Missing Secrets Error. Then adding missing secrets and deploying.

 **Link to the latest (third) system test that I run manually on Jenkins: (They passed)**
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1661/pipeline

### Manual tests
1. I tested a blueprint with one missing secret.
2. I tested a blueprint with two missing secrets.
3. I tested a blueprint that requires two secrets and I added only one of them in the resources/secrets screen. I made sure that only one secret is missing and that I can add it through the deploy modal.

<!--- If applicable, describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If applicable, provide a link to system tests, e.g: -->
<!--- [#1039](https://jenkins.cloudify.co/.../1039) (2021-08-06 07:05:50) --> 
<!--- If not applicable, just write “N/A” in this section. -->

## Documentation
N/A
<!--- If applicable, describe how you documented or plan to document your changes. -->
<!--- Provide links to JIRA issues, other PRs or related documentation pages. -->
<!--- If not applicable, just write “N/A” in this section. -->

qooban commented on widgets/common/src/GenericDeployModal.tsx:
```diff
@@ -243,6 +269,55 @@ class GenericDeployModal extends React.Component {
+    parseMissingSecrets() {
+        const { errors } = this.state;
+        return errors.error
+            .replace('dsl_parser.exceptions.UnknownSecretError: Required secrets: [', '')
+            .replace("] don't exist in this tenant", '')
+            .split(',');
```
Have you thought about using regular expression to get just "['secretA', 'secretB', ...]" part? 

Assuming that "dsl_parser.exceptions.UnknownSecretError" has always the same syntax, it should be possible to make it smarter and reuse "dsl_parser.exceptions.UnknownSecretError" between `checkMissingSecretsError` and `parseMissingSecrets`. WDYT?

qooban commented on widgets/common/src/GenericDeployModal.tsx:
```diff
@@ -305,7 +382,12 @@ class GenericDeployModal extends React.Component {
-                    <Form errors={errors} scrollToError onErrorsDismiss={() => this.setState({ errors: {} })}>
+                    <Form
+                        errors={isMissingSecrets ? this.missingSecretsError() : errors}
+                        errorMessageHeader={isMissingSecrets ? 'Missing Secrets Error' : undefined}
```
If you'd like to change the header, let's do it with translation support.

I'm not sure we need to explicitly change the header.

qooban commented on package.json:
```diff
@@ -56,7 +56,7 @@
     "cloudify-blueprint-topology": "^2.5.3",
     "cloudify-ui-common": "^4.1.0",
-    "cloudify-ui-components": "^2.9.0",
+    "cloudify-ui-components": "file:../cloudify-ui-components/cloudify-ui-components-2.10.0.tgz",
```
I assume this will be changed once you release new version of `cloudify-ui-components`.

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,90 @@
+                <Form errors={errors} onErrorsDismiss={clearErrors}>
+                    {secretKeysArr.map(field => (
+                        <UnsafelyTypedFormField key={field}>
+                            <Input type="text" name={field} label={field} onChange={setSecretInputs} />
```
1. Shouldn't we use `type="password"`?
    In RD-1607 there is:
    > The inputs should be of type password so the values won’t appear in clear text.

2. Why not using `Form.Input`? TypeScript issues?

3. Shouldn't we pass `required` prop to indicate it is mandatory (as you already require it `onSave` method)?

leon-good-life replied:
When using `Form.Field` I'm getting:
```
Type '{ children: Element; key: string; required: true; label: string; }' is not assignable to type 'IntrinsicAttributes & object'.
  Property 'children' does not exist on type 'IntrinsicAttributes & object'.ts(2322)
```

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,90 @@
+        const keys = Object.keys(secretInputs);
+
+        if (keys.some(key => _.isEmpty(secretInputs[key]))) {
+            setErrors({});
```
You can probably use `clearErrors` here.

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,90 @@
+    const onSave = () => {
+        const keys = Object.keys(secretInputs);
+
+        if (keys.some(key => _.isEmpty(secretInputs[key]))) {
```
`isEmpty` according to https://lodash.com/docs/4.17.15#isEmpty is used to check "if value is an empty object, collection, map, or set". As `secretInputs[key]` is supposed to be of type string I would use another method here - either check if `secretInputs[key]` is falsy or check it's length.

qooban reviewed (commented):
Added some comments. 

Not able to test UI yet (without `cloudify-ui-components` updated).

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,90 @@
+                    onClick={onSave}
+                    disabled={isLoading}
+                    content={t('buttons.save')}
+                    icon="cogs"
```
Let's use proper icon here, maybe plus icon?

leon-good-life commented:
I'm working on fixing saving multiple secrets.

qooban commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,51 @@
+                toolbox={toolbox}
+                secretKeys={parseMissingSecrets()}
+                open={secretsModalVisible}
+                onClose={hideSecretsModal}
+                resolveError={resolveError}
```
`resolveError` prop name is misleading. It is making `SecretsModal` being wired to the current use-case. I think if you really want to distinguish between different actions in the modal, I would rather name it `onSave`, `onAdd` , `onCloseWithSave` or `onCloseWithAdd` (depending on the action/button name).

qooban commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,51 @@
+            </List>
+            <SecretsModal
+                toolbox={toolbox}
+                secretKeys={parseMissingSecrets()}
```
L28 raises a question: what about just passing `secretKeys` here?

qooban commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,51 @@
+    return (
+        <>
+            <Button floated="right" style={missingSecretsButtonStyle} onClick={showSecretsModal}>
+                Add missing secrets
+            </Button>
+            <p style={{ display: 'inline' }}>The following required secrets are missing in this tenant:</p>
```
We need translation here.

leon-good-life replied:
I added

qooban commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,51 @@
+interface props {
+    error: string;
+    toolbox: Stage.Types.Toolbox;
+    resolveError: () => void;
```
Probably to follow naming conventions it would be better to name it starting with `on`.

qooban commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,51 @@
+import type { FunctionComponent } from 'react';
+import SecretsModal from './SecretsModal';
+
+const missingSecretsButtonStyle = {
+    marginRight: 10
+};
```
Optional:

Seems like it is used only in one place, so maybe not worth to create a const for it?

leon-good-life replied:
If I'll make an inline style `style={ { marginRight: 10 } }`, then `MissingSecretsButton` will re-render every time `MissingSecretsError` re-renders. Thats because new instance of an object passed as a `style` prop, and React re-renders components when props are changing. Just re-rendering this button maybe will not affect much the performance, but the habit of writing inline styles generally will do affect the performance in my opinion.

leon-good-life replied:
Because you noted that your comment here is optional, I guess I can click on resolve button.

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,89 @@
+    return (
+        <Modal open={open} onClose={onClose}>
+            <Modal.Header>
+                <Icon name="key" />
```
Looking at what icons we use in connection with secrets it seems like `plus` icon would be better.

See icon for Secrets page and icons used in Secrets Store Management widget.

leon-good-life replied:
I changed it as you suggested.

qooban commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,51 @@
+    const { useBoolean } = Stage.Hooks;
+    const [secretsModalVisible, showSecretsModal, hideSecretsModal] = useBoolean(false);
+
+    function parseMissingSecrets() {
+        // Get comma separated values inside square brackets
+        const matches = error.match(/\[(.*?)\]/);
+        if (matches && matches.length > 1) {
+            return matches[1].split(', ');
+        }
+        return [];
+    }
+
+    const { Button, List } = Stage.Basic;
+    const secretKeys = parseMissingSecrets();
```
I see that `parseMissingSecrets` can be used only by `secretKeys` (if you address another comment). 

What do you think about having:
```
const secretKeys = useMemo(() => /* the whole logic for parsing missing secrets */, [error]);
```
then?

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,89 @@
+            })
+        )
+            .catch(setMessageAsError)
+            .finally(() => {
```
Did you check if we need also call `clearErrors` in `finally` step?

leon-good-life replied:
Yes, when opening SecretsModal, leaving the inputs empty, clicking on "Add" button, there are errors, closing the modal and opening it again, the errors are still there.

I added `clearErrors`.

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,89 @@
+            <Modal.Content>
+                <Form errors={errors} onErrorsDismiss={clearErrors}>
+                    {secretKeys.map(field => (
+                        <Form.Input key={field}>
+                            <Input type="password" required name={field} label={field} onChange={setSecretInputs} />
+                        </Form.Input>
```
Can you elaborate why you didn't use 
```
<Form.Field ...>
  <Form.Input ... />
</Form.Field ...>
```
format as in other places?

leon-good-life replied:
I changed it to this format. But I'm using `UnsafelyTypedFormField` because whenI'm  using `Form.Field` I'm getting:
```
Type '{ children: Element; key: string; required: true; label: string; }' is not assignable to type 'IntrinsicAttributes & object'.
  Property 'children' does not exist on type 'IntrinsicAttributes & object'.ts(2322)
```

qooban reviewed (commented):
Added some more comments.

leon-good-life commented on app/translations/en.json:
```diff
@@ -539,6 +539,15 @@
+                        "save": "save"
+                    },
+                    "errors": {
+                        "noSecretValues": "Please provide the secret values"
```
I changed it as you suggested.

leon-good-life commented on widgets/common/src/GenericDeployModal.tsx:
```diff
@@ -12,6 +13,7 @@ class GenericDeployModal extends React.Component {
         deploymentInputs: [],
         deploymentName: '',
         errors: {},
+        isMissingSecrets: false,
```
I changed it as you suggested.

leon-good-life commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,89 @@
+        const visibility = defaultVisibility;
+        const actions = new Stage.Common.SecretActions(toolbox);
+        Promise.all(
+            keys.map(secretKey => {
+                return actions.doCreate(secretKey, secretInputs[secretKey], visibility, isHiddenValue);
+            })
```
I changed the fields, now the are with a label above:
<img width="948" alt="Screen Shot 2021-12-08 at 14 37 04" src="https://user-images.githubusercontent.com/17104212/145209503-8f2a72ab-0cbd-4aa3-86e0-63562809f936.png">

leon-good-life replied:
I changed as you suggested.

leon-good-life commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,90 @@
+            </Modal.Header>
+
+            <Modal.Content>
+                <Form errors={errors} onErrorsDismiss={clearErrors}>
```
I changed the fields, now the are with a label above:
<img width="948" alt="Screen Shot 2021-12-08 at 14 37 04" src="https://user-images.githubusercontent.com/17104212/145209503-8f2a72ab-0cbd-4aa3-86e0-63562809f936.png">

leon-good-life replied:
We agreed to skip it for now.

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,88 @@
+
+        const isHiddenValue = true;
+        const visibility = defaultVisibility;
+        const actions = new Stage.Common.SecretActions(toolbox);
```
General:

I see that there are uses of of `Stage.Common` global inside `widgets/common/src`. I'm not sure it is the right way to go. I think you should import `SecretsActions` file directly instead of accessing it with `Stage.Common` global.

This is general comment. If it works, then let's leave it, but maybe we should change that in the future.

leon-good-life replied:
I tried a quick fix, then decided to leave it as it because I understand that it is optional.

qooban commented on widgets/common/src/SecretActions.ts:
```diff
@@ -32,6 +32,11 @@ class SecretActions {
 }
+declare global {
+    namespace Stage.Common {
+        export default { SecretActions };
```
I'm getting error when starting webpack dev server:
```
ERROR in widgets/common/src/SecretsModal.tsx:44:42
TS2339: Property 'SecretActions' does not exist on type 'typeof Common'.
    42 |         const isHiddenValue = true;
    43 |         const visibility = defaultVisibility;
  > 44 |         const actions = new Stage.Common.SecretActions(toolbox);
       |                                          ^^^^^^^^^^^^^
    45 |         Promise.all(
    46 |             keys.map(secretKey => actions.doCreate(secretKey, secretInputs[secretKey], visibility, isHiddenValue))
    47 |         )
```

I think you need to export `SecretActions` class properly (see L2).

BTW I suggest to migrate fully this file to TS (the class seems very simple and obvious to add typings).

leon-good-life replied:
Fixed. Migrated `SecretActions.ts` to TypeScript. Added `Visibility` type.

qooban reviewed (commented):
Tested UI finally, looks good. Added more comments to the code.

I'm not sure we should use password type for all type of secrets (we don't do that in Secrets widget and we don't do that for some secrets in Getting Started modal), but that's out of scope of this ticket and PR.

leon-good-life commented on widgets/common/src/GenericDeployModal.tsx:
```diff
@@ -282,6 +296,8 @@ class GenericDeployModal extends React.Component {
             deploymentName,
             errors,
+            areSecretsMissing,
+            secretsModalVisible,
```
I forgot to remove it. I pushed a "Fix PR" commit that removes it.

leon-good-life commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,52 @@
+    return (
+        <>
+            <Button floated="right" style={missingSecretsButtonStyle} onClick={showSecretsModal}>
+                Add missing secrets
```
I pushed a "Fix PR" commit that adds a translation.

leon-good-life commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,52 @@
+const { i18n } = Stage;
+const t = (key: string) => i18n.t(`widgets.common.deployments.deployModal.${key}`);
+
+const missingSecretsError: FunctionComponent<props> = ({ error, toolbox, onAdd }) => {
```
I named the component starting with a lowercase letter by mistake. 
I pushed a "Fix PR" commit that changes it to capital letter.

leon-good-life commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,52 @@
+import { FunctionComponent, useMemo } from 'react';
```
Changed it.

leon-good-life commented on widgets/common/src/GenericDeployModal.tsx:
```diff
@@ -120,8 +130,12 @@ class GenericDeployModal extends React.Component {
             });
         });
 
+        const checkMissingSecretsError = errors => {
```
Changed it.

leon-good-life commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -128,6 +128,18 @@ describe('Blueprints widget', () => {
+            cy.contains('.modal button', 'Deploy').click();
+            cy.get('form.error .error .header').should('have.text', 'Missing Secrets Error');
+        });
+    });
```
1. I extended the test.
2. I changed to using only `it` section

leon-good-life commented:
Only one test is failing in Jenkins system tests.
I run this test locally, and it failed also locally.
The failed test is: "Blueprints widget - for specific blueprint - should open Composer with the blueprint imported on "Edit a copy in Composer" icon click"
I'm working on fixing it.

Uploading Screen Recording 2021-12-09 at 21.48.59.mov…

NorbertAmartus commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,53 @@
+    const { Button, List } = Stage.Basic;
+    const secretKeys = useMemo(() => {
+        // Get comma separated values inside square brackets
+        const matches = error.match(/\[(.*?)\]/);
```
I may have an suggestion on creating a new variable
```typescript
const comaSeparatedValuesRegexp = /\[(.*?)\]/;
const matches = error.match(comaSeparatedValuesRegexp);
```

To not create a comment, when we don't need it, cause the code would be self explanatory
But it's an optional suggestion :)

Let me know what do you think about it 😉

leon-good-life replied:
I removed the comment and I moved it to a function a new comaSeparatedValuesInBrackets: https://github.com/cloudify-cosmo/cloudify-stage/pull/1693/commits/d79e94241e7721bb1a11873f1e4d2c7ddd29f87b

leon-good-life commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,53 @@
+}
+
+const { i18n } = Stage;
+const t = (key: string) => i18n.t(`widgets.common.deployments.deployModal.${key}`);
```
Thank you, after your review, I pushed a commit that changes that.

leon-good-life commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,87 @@
+};
+
+const { i18n } = Stage;
+const t = (key: string) => i18n.t(`widgets.common.deployments.secretsModal.${key}`);
```
I changed that as you suggested.

leon-good-life commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,53 @@
+    const secretKeys = useMemo(() => {
+        // Get comma separated values inside square brackets
+        const matches = error.match(/\[(.*?)\]/);
+        if (matches && matches.length > 1) {
```
Sometimes TypeScript doesn't like optional chaining, I guess in this case, it is because TypeScript expects a number to compare it with 1.
<img width="478" alt="Screen Shot 2021-12-10 at 10 49 01" src="https://user-images.githubusercontent.com/17104212/145544977-dceccebf-726c-49be-ae6d-91ac067d63b7.png">

Probably if I will rewrite my regex I will not need this `if` and `split(', ')`

NorbertAmartus replied:
Thanks for checking that out!
In that case I think that it would be best to leave it as it was :)

I'll resolve this discussion, thanks for taking a look on that!

NorbertAmartus reviewed (approved):
Thanks for addressing all the issues!
Looks neat 😎

kubama reviewed (commented):
Looks like there are already two robust reviews done, so I'll opt-out from this one.

leon-good-life commented:
I fixed the failing system test (it caused by tests structure - when the test of missing error message was on the top of the file it caused the test below it to fail). Here is the link to the Jenkins: 
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1654/pipeline 
(This comment was written before Jenkins system tests finished running, but I'm pretty sure that they will pass because they passed on my computer)

I worked almost all the day today on trying to fix the exceeding of the maximum bundle size limit e.g. `npm run size`. The following commit: 
Merge branch 'master' of https://github.com/cloudify-cosmo/cloudify-s… 93589e7 caused the Main JS Bundle to exceed the max build size. I don't have any idea why Vendor JS bundle exceeded. But during my deep investigation today I was realized about `.size-limit` file that the size limits configured in it. This file is updated several times a month - and the limit is increasing in each commit to that file. I asked in slack when is it okay to update that file.

leon-good-life commented:
System tests passed:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1654/pipeline/

qooban commented on widgets/common/src/MissingSecretsError.tsx:
```diff
@@ -0,0 +1,53 @@
+    const [secretsModalVisible, showSecretsModal, hideSecretsModal] = useBoolean(false);
+
+    const { Button, List } = Stage.Basic;
+    const secretKeys = useMemo(() => comaSeparatedValuesInBrackets(error), [error]);
```
I'm wondering why you decided to extract `comaSeparatedValuesInBrackets` function?

If the reason was to name the action, then IMHO it is misleading. I would expect it to have verb inside. Also it does not produce comma separated values, it converts/parses a string with comma-separated values in brackets into an array.

If the reason was to reuse it in the future - I don't think it is the case - it is used only once and it seems not probable it will be used anywhere else.

leon-good-life replied:
The reason was to remove the comment I had before and to make the code self explanatory without using a comment.

leon-good-life replied:
> I removed the comment and I moved it to a function a new comaSeparatedValuesInBrackets: [d79e942](https://github.com/cloudify-cosmo/cloudify-stage/commit/d79e94241e7721bb1a11873f1e4d2c7ddd29f87b)

leon-good-life replied:
I refactored to `parseCommaSeparatedValuesInBrackets` as agreed in our conversation. Commit: 28b4b8c4edf900ac4932289460a5f71c2aa6af94

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -514,4 +514,32 @@ describe('Blueprints widget', () => {
+        const secretName = 'test';
+        const blueprintName = 'blueprint_with_required_secrets';
+        before(() => cy.uploadBlueprint('blueprints/required_secrets.zip', blueprintName).refreshPage());
+        after(() => cy.stageRequest(`/console/sp/secrets/${secretName}`, 'DELETE'));
```
I think we shouldn't use `after` to remove secret.
Ref.: https://docs.cypress.io/guides/references/best-practices#Using-after-or-afterEach-hooks

See other tests how we prepare test to start with desired state.

leon-good-life replied:
I removed this `after`

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -514,4 +514,32 @@ describe('Blueprints widget', () => {
+            getBlueprintRow(blueprintName).find('.rocket').click();
+            cy.get('input[name=deploymentName]').type('blahBlahBlah');
+            cy.contains('.modal button', 'Deploy').click();
+            cy.get('form.error .error .header').should('have.text', 'Missing Secrets Error');
+            cy.get('form.error .error p').should(
+                'have.text',
+                'The following required secrets are missing in this tenant:'
+            );
+            cy.get('form.error .error .item').should('have.text', secretName);
```
You can probably use `.within` to make it a bit more structured:
```suggestion
            cy.get('form.error .error').within(() => {
              cy.get('.header').should('have.text', 'Missing Secrets Error');
              cy.get('p').should('have.text', 'The following required secrets are missing in this tenant:');
              cy.get('.item').should('have.text', secretName);
            });
```

leon-good-life replied:
I changed to using `.within`

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -514,4 +514,32 @@ describe('Blueprints widget', () => {
+        before(() => cy.uploadBlueprint('blueprints/required_secrets.zip', blueprintName).refreshPage());
+        after(() => cy.stageRequest(`/console/sp/secrets/${secretName}`, 'DELETE'));
+
+        it('should open Missing Secrets Error message when deploying blueprint with missing secrets', () => {
+            getBlueprintRow(blueprintName).find('.rocket').click();
+            cy.get('input[name=deploymentName]').type('blahBlahBlah');
+            cy.contains('.modal button', 'Deploy').click();
+            cy.get('form.error .error .header').should('have.text', 'Missing Secrets Error');
+            cy.get('form.error .error p').should(
+                'have.text',
+                'The following required secrets are missing in this tenant:'
+            );
+            cy.get('form.error .error .item').should('have.text', secretName);
+            cy.contains('form.error button', 'Add Missing Secrets').click();
+            cy.contains('.secretsModal button', 'Add').click();
+            cy.get('.secretsModal .error .header').should('have.text', 'Errors in the form');
+            cy.get('.secretsModal .error li').should('have.text', 'Please provide values for secrets');
+            cy.get('.secretsModal input').type('aaa');
+            cy.contains('.secretsModal button', 'Add').click();
+            cy.get('form.error .error').should('not.exist');
+            cy.contains('.modal button', 'Deploy').click();
+            cy.get('.modal').should('not.exist');
+        });
```
The code is very dense. I suggest to add more empty lines to make it easier to see logical part (action - assertion - action - assertion - ...).

leon-good-life replied:
I added empty lines after assertions

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -514,4 +514,32 @@ describe('Blueprints widget', () => {
+            cy.get('form.error .error .item').should('have.text', secretName);
+            cy.contains('form.error button', 'Add Missing Secrets').click();
+            cy.contains('.secretsModal button', 'Add').click();
+            cy.get('.secretsModal .error .header').should('have.text', 'Errors in the form');
+            cy.get('.secretsModal .error li').should('have.text', 'Please provide values for secrets');
+            cy.get('.secretsModal input').type('aaa');
```
You can use similar patter with `.within` here.

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -514,4 +514,32 @@ describe('Blueprints widget', () => {
+            cy.contains('.secretsModal button', 'Add').click();
+            cy.get('form.error .error').should('not.exist');
+            cy.contains('.modal button', 'Deploy').click();
+            cy.get('.modal').should('not.exist');
```
This is a bit risky as it might take some time for modal to be closed. Take a look in other places how it is solved. I suggest to

leon-good-life replied:
The default timeout is 4 seconds and it was working every time I run the test, I can change it to `cy.get('.modal', { timeout: 6000 }).should('not.exist');` to increase to 6 seconds. What do you think?

leon-good-life replied:
In our meeting this morning, You suggested to remove the following lines:

```
cy.contains('.modal button', 'Deploy').click();
cy.get('.modal').should('not.exist');
```
So, I will remove them.

leon-good-life replied:
And to check the request instead.

leon-good-life replied:
I changed it to `interceptSp`.

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -514,4 +514,32 @@ describe('Blueprints widget', () => {
+
+    describe('for blueprint with required secrets', () => {
+        const secretName = 'test';
+        const blueprintName = 'blueprint_with_required_secrets';
```
Use `blueprintNamePrefix` in `blueprintName` to make it consistent with the rest of the file.

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -514,4 +514,32 @@ describe('Blueprints widget', () => {
+    describe('for blueprint with required secrets', () => {
+        const secretName = 'test';
+        const blueprintName = 'blueprint_with_required_secrets';
+        before(() => cy.uploadBlueprint('blueprints/required_secrets.zip', blueprintName).refreshPage());
```
The test will fail when 'blueprint_with_required_secrets' blueprint is already uploaded. Check how `blueprintNamePrefix` in main `before` section is used to overcome this issue.

leon-good-life replied:
Fixed that.

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -514,4 +514,32 @@ describe('Blueprints widget', () => {
         });
     });
+
+    describe('for blueprint with required secrets', () => {
```
In https://github.com/cloudify-cosmo/cloudify-stage/pull/1693#discussion_r765686270 you mentioned that you "changed to using only it section", but I still see - `describe` and `it`.

leon-good-life replied:
I changed it back because I'm using `before` and `after`, and to be more consistent with the rest of the file.

leon-good-life replied:
I moved the test to `deployment_button_spec.ts` and changed to `it`.

qooban commented on app/utils/StageAPI.ts:
```diff
@@ -297,6 +297,8 @@ declare global {
 
             type ObjectKeys<T extends Record<string, any>> = T[keyof T];
+
+            type Visibility = 'private' | 'tenant' | 'global';
```
This type is not fitting here IMHO. It's not Stage-related. It is a type describing allowed values for a parameter provided by Cloudify Manager API.

Until we have RD-1874 implemented, I don't have good suggestion for a place to store it. 
For now I would add it to the place where you actually need it - `SecretActions`?

leon-good-life replied:
moved.

qooban reviewed (commented):
The code mostly looks good 👍🏻

Added some comments. Most of them are related to the new system test.

leon-good-life commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,86 @@
+        setLoading();
+
+        const isHiddenValue = true;
+        const visibility = Stage.Common.Consts.defaultVisibility as Stage.Types.Visibility;
```
I'm not to sure where to put the type only to `defaultVisibility` without making too much refactoring:
<img width="545" alt="Screen Shot 2021-12-14 at 12 22 58" src="https://user-images.githubusercontent.com/17104212/145980139-12f26bdb-f3a3-4e00-abcf-dfaf2f542405.png">

We agreed to leave it as is in our conversation this morning.

leon-good-life commented:
I started system tests after moving the system test to deploymentButtonSpec and addressing PR comments there.
Link to Jenkins system tests: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1661/pipeline

qooban commented on test/cypress/integration/widgets/deployment_button_spec.ts:
```diff
@@ -226,6 +236,32 @@ describe('Create Deployment Button widget', () => {
+            cy.contains('.secretsModal button', 'Add').click();
+            cy.get('form.error .error').should('not.exist');
+
+            cy.interceptSp('PUT', `/console/sp/secrets/${secretName}`, { statusCode: 200 });
```
This line will have no effect.

You need to call `interceptSp` before the specific HTTP request is expected to be called. In a place where it is expected you need to use `cy.wait`.

See e.g. 'parses constraint error message from /deployments REST API' test case for how to wait for HTTP request (there's `deployBlueprint` awaited).

leon-good-life replied:
Fixed. I added `.as('addSecrets')` to  and `cy.wait('@addSecrets');` in the following commit: 6ed00b3

qooban commented on test/cypress/integration/widgets/deployment_button_spec.ts:
```diff
@@ -226,6 +236,32 @@ describe('Create Deployment Button widget', () => {
+            cy.get('.secretsModal .error .header').should('have.text', 'Errors in the form');
+            cy.get('.secretsModal .error li').should('have.text', 'Please provide values for secrets');
+
+            cy.get('.secretsModal input').type('aaa');
```
Will
```suggestion
            cy.getField(secretName).type('aaa');
```
work here?

leon-good-life replied:
<img width="1432" alt="Screen Shot 2021-12-14 at 16 49 35" src="https://user-images.githubusercontent.com/17104212/146021342-e0ebf3c8-f4dc-4349-8323-cbf2163868c1.png">

qooban replied:
My bad, the right solution is to use `cy.getField(secretName).find('input').type('aaa')`.

qooban commented on widgets/common/src/SecretActions.ts:
```diff
@@ -1,37 +1,43 @@
-// @ts-nocheck File not migrated fully to TS
-export {};
+type Visibility = 'private' | 'tenant' | 'global';
```
Can you export it from here and reuse it in `SecretsModal`?

qooban commented on widgets/common/src/SecretsModal.tsx:
```diff
@@ -0,0 +1,88 @@
+import type { FunctionComponent } from 'react';
+
+type Visibility = 'private' | 'tenant' | 'global';
```
You can import that type from `SecretActions` I believe.

leon-good-life replied:
I made a commit: "Reuse type Visibility in SecretActions and SecretsModal files 5a0aff7"

qooban commented on test/cypress/integration/widgets/deployment_button_spec.ts:
```diff
@@ -145,7 +155,7 @@ describe('Create Deployment Button widget', () => {
 
     describe('handles errors during deploy & install process', () => {
         afterEach(() => {
-            cy.get(`.actions > .ui:nth-child(1)`).click();
```
I tested it locally and I think you don't need any changes here and adding `clickIfExist` function seems not necessary.

It just closes the modal after each test.

leon-good-life replied:
Fixed. I reverted the clickIfExist that I introduced. I run the tests and they passed.

qooban reviewed (commented):
Added few more comments.
Please revert `package-lock.json` file.

leon-good-life commented:
I run system tests again: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1664/pipeline

qooban commented on test/cypress/integration/widgets/deployment_button_spec.ts:
```diff
@@ -226,6 +228,32 @@ describe('Create Deployment Button widget', () => {
+            cy.get('.secretsModal .error li').should('have.text', 'Please provide values for secrets');
+
+            cy.get('.secretsModal input').type('aaa');
+            cy.interceptSp('PUT', `/secrets/${secretName}`, { statusCode: 200 }).as('addSecrets');
```
The 3rd argument will work like a static response, so Cypress will intercept this request and always respond with statusCode = 200.

See `interceptSp` definition + see `cy.intercept` documentation (https://docs.cypress.io/api/commands/intercept#Syntax)

I think you should have here:
```suggestion
            cy.interceptSp('PUT', `/secrets/${secretName}`).as('addSecrets');
```
That way Cypress will just listen for `PUT /secrets/${secretName}` requests and won't change anything inside.

leon-good-life replied:
I fixed in the following PR: https://github.com/cloudify-cosmo/cloudify-stage/pull/1705

NorbertAmartus commented:
@leon-good-life 

> I run system tests again: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1664/pipeline

It seems that there was some sort of network problem
I've executed tests again, so that we'd have a fresh view on them 😉: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1668/pipeline

qooban commented on test/cypress/integration/widgets/deployment_button_spec.ts:
```diff
@@ -226,6 +228,32 @@ describe('Create Deployment Button widget', () => {
             cy.get('@string_constraint_pattern').should('have.class', 'error');
         });
+
+        it('should open Missing Secrets Error message when deploying blueprint with missing secrets', () => {
```
What do you think about having that test in the following shape?:
```
       it('should handle missing secrets error', () => {
            const secretName = 'test';
            cy.deleteSecrets(secretName);

            selectBlueprintInModal('required_secrets');
            cy.get('.modal').within(() => {
                cy.getField('Deployment name').find('input').type('blahBlahBlah');
                cy.contains('button', 'Deploy').click();
                cy.get('.error.message').within(() => {
                    cy.get('.header').should('have.text', 'Missing Secrets Error');
                    cy.get('p').should('have.text', 'The following required secrets are missing in this tenant:');
                    cy.get('.item').should('have.text', secretName);
                });

                cy.contains('button', 'Add Missing Secrets').click();
            });

            cy.get('.secretsModal').within(() => {
                cy.interceptSp('PUT', `/secrets/${secretName}`).as('addSecrets');

                cy.contains('button', 'Add').click();
                cy.get('.error.message').within(() => {
                    cy.get('.header').should('have.text', 'Errors in the form');
                    cy.get('li').should('have.text', 'Please provide values for secrets');
                });

                cy.getField(secretName).find('input').type('aaa');
                cy.contains('button', 'Add').click();
                cy.wait('@addSecrets');
            });

            cy.get('.error.message').should('not.exist');
        });
```
Let's discuss it in our meeting today.

leon-good-life replied:
I refactored and fixed in the following PR: https://github.com/cloudify-cosmo/cloudify-stage/pull/1705

qooban reviewed (commented):
The only thing to fix is system test. Let's discuss in our meeting and finalize it.

qooban commented:
@leon-good-life Wait, wait, I asked to discuss system test fixes in our meeting.

Also, didn't we agree on having a rule that either @kubama or me should approve PR before merging to master?
Anyways, it's not good to merge if there are still open discussions.

leon-good-life commented:
Sorry for that, I merged by mistake, I'm fixing right now. Is it possible to revert my merge?

leon-good-life commented:
The fixes will be in a new PR

leon-good-life commented:
Here is the follow up PR: https://github.com/cloudify-cosmo/cloudify-stage/pull/1705
